### PR TITLE
Dockerfile: arm64 support

### PR DESCRIPTION
**Pull request by thepwagner**
This PR amends the `Dockerfile` and `bin/docker-dev-shell` commands to support native builds on arm64, as provided by an M1 Mac.

* Initial commits walk the Dockerfile and provide arm64-native alternatives when binaries are fetched.
   * There is no Elm, sorry.
   * There are no Terraform Debian packages, so I switched to fetching zips.
   * In places where the digest of downloaded files is verified, you have two digests to maintain now.
* No attempt is made to build `arm64` in CI, or publish `arm64` releases.

The resulting build stands up OK to light testing: I could dry-run a golang, node and gradle repository.

I hit an error in the bundler that I didn't try to debug:
```
	 1: from /home/dependabot/dependabot-core/bundler/lib/dependabot/bundler/update_checker/shared_bundler_helpers.rb:54:in `block in in_a_native_bundler_context'
/home/dependabot/dependabot-core/bundler/lib/dependabot/bundler/update_checker/conflicting_dependency_resolver.rb:44:in `block in conflicting_dependencies': undefined method `name' for nil:NilClass (NoMethodError)
```

### CODE REVIEW

**Comment by jeffwidman** [comment]
I opened a PR a while back for the `focal` upgrade: https://github.com/dependabot/dependabot-core/pull/4394

Although at this point `jammy` gets released next month, so if it's hard for the core team to pull in [due to scheduling](https://github.com/dependabot/dependabot-core/pull/4394#issuecomment-969275077), they may want to jump straight to that--better to test one release well than to test two releases less rigorously.

**Comment by jurre** [comment]
Thanks @thepwagner! Bumping ubuntu seems reasonable to me, although I'd like to roll that out separately, which I might as well do now, and so will.

@jeffwidman yeah it's not a bad call, although I kind of don't want to wait even longer, and we should just prioritize rolling that out when it's released. We have a little bit more slack on the team now so we should be able to prioritize this type of work quicker

File: Dockerfile
Comment (mctofu) [inline Dockerfile:263]: I'm curious about these changes since they seem unrelated to arm support? I think this ends up getting deleted anyway below in the `find ... -delete` command?

Comment (pavera) [reply Dockerfile:263]: Not an expert, but it appears flutter ships with a dart-sdk, these 2 lines replace the x86 dart sdk with the arm sdk downloaded in the previous step.  I say this because without these 2 lines I get the following error during docker build:
`#26 47.08 qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory`

Comment (thepwagner) [reply Dockerfile:263]: Everyone is correct 🎉 :

The flutter SDK (even after https://github.com/dependabot/dependabot-core/pull/4904/files), embeds an `x86_64` dart sdk:

```
$ file /opt/flutter/new/flutter/bin/cache/dart-sdk/bin/dart 
/opt/flutter/new/flutter/bin/cache/dart-sdk/bin/dart: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, stripped
```

And the files are deleted by the find, so they aren't part of the final image.

IIUC the issue is L266 where `flutter --version` is called before the delete happens - that uses the bundled `dart-sdk` and fails, so I ninja-patched the architecture-appropriate `dart-sdk` in.

If `flutter --version` is just meant to "hello world"/verify the installation, removing that step is an alternative to this hack.

Comment (mctofu) [reply Dockerfile:263]: Ah, got it, thanks! I don't think we need to make it conditional, I was just curious why it was necessary. This is worth adding a comment about?

cc: @jonassf @sigurdm to see if you have any feedback here.

Comment (sigurdm) [reply Dockerfile:263]: > If flutter --version is just meant to "hello world"/verify the installation, removing that step is an alternative to this hack.

I think so! When you clone flutter from github the first invocation will create the `version` file (that pub relies on) - that is why we did this step, but the packaged flutter contains it already - so I think it can just be removed!

**Review by mctofu — APPROVED** [review]
Thanks!

**Comment by jurre** [comment]
Just running it on my old, old non-arm64 laptop to make sure that still works, but assume it will after which I'll merge this in. Thanks @thepwagner!

**Comment by jurre** [comment]
Hmm, running into a little snag:

```
 => ERROR [stage-0 13/38] RUN cd /tmp   && curl --http1.1 -o go-x86_64.tar.gz https://dl.google.com/go/go1.18.linux-x86_64.tar.gz   && printf "e85278e98f57cdb150fe8409e6e5df5343ecb13cebf03a5d5ff12bd55a80264f go-amd64.tar.gz\n7ac7b396a691e588c5fb57  0.6s
------
 > [stage-0 13/38] RUN cd /tmp   && curl --http1.1 -o go-x86_64.tar.gz https://dl.google.com/go/go1.18.linux-x86_64.tar.gz   && printf "e85278e98f57cdb150fe8409e6e5df5343ecb13cebf03a5d5ff12bd55a80264f go-amd64.tar.gz\n7ac7b396a691e588c5fb57687759e6c4db84a2a3bbebb0765f4b38e5b1c5b00e go-arm64.tar.gz\n" | sha256sum -c --ignore-missing -   && tar -xzf go-x86_64.tar.gz -C /opt   && rm go-x86_64.tar.gz:
#19 0.344   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
#19 0.344                                  Dload  Upload   Total   Spent    Left  Speed
100  1449  100  1449    0     0   7748      0 --:--:-- --:--:-- --:--:--  7748
#19 0.537 sha256sum: 'standard input': no file was verified
------
```

I can dig into that later

**Comment by thepwagner** [comment]
@jurre I think the issue is with how `bin/docker-dev-shell` is setting `TARGET_ARCH`.

From `-o go-x86_64.tar.gz` in the output, we can reason `TARGETARCH=x86_64` for you, but from `printf "... go-amd64.tar.gz"`, I expected `TARGETARCH=amd64`.

In CI, the `bin/docker-dev-shell` script isn't executed - so it uses the built-in default of `amd64` and works great.
If `uname -m` returns `x86_64` for you, then we probably just need a little extra script in `docker-dev-shell` to map that to `amd64`.

I'm confident that `TARGET_ARCH=arm64` works as expected on M1:
```
$ uname -a
Darwin trolololol.local 21.4.0 Darwin Kernel Version 21.4.0: Mon Feb 21 20:35:58 PST 2022; root:xnu-8020.101.4~2/RELEASE_ARM64_T6000 arm64
$ uname -m
arm64
```

**Comment by pavera** [comment]
Yes @thepwagner I also have confirmed that the same error occurs on amd64 and arm64.  The dry run script for bundler expects/requires a lock file in the repo used for testing.  I've also confirmed that bundler works on both amd64 and arm64 as long as the repo has a lock file.

**Comment by jurre** [comment]
image

**Pull request merged**
